### PR TITLE
pgbackrest as a postgresql backup solution

### PR DESCRIPTION
**Pull request by javierwilson**
this doesn't take full advantage of pgbackrest, specially the "backup from standby" option :( we could benefit from it is we use pgbackrest for the standby configuration itself, this would also automate standby installations more easily.  tested only on dev enviroment so far.

### CODE REVIEW

File: ansible/roles/pgbackrest/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    owner=postgres
+    group=postgres
+    mode=640
+
+- name: Creating postgres stanza
+  become: True
+  become_user: postgres
+  vars:
+    ansible_ssh_pipelining: True
+  shell: pgbackrest --stanza=backup --log-level-console=info stanza-create
```
Comment (snopoke) [inline ansible/roles/pgbackrest/tasks/main.yml]: Is this OK to run multiple times?

Comment (javierwilson) [reply ansible/roles/pgbackrest/tasks/main.yml]: it is, however it is marked as "changed" everytime,  i am going to check if {{postgresql_backup_dir}}/archive/backup exists before running it again.

Comment (javierwilson) [reply ansible/roles/pgbackrest/tasks/main.yml]: done. however i still have the problem of postgresql restarting twice because of the changes done to the archive_command line by the two different playbooks (postgresql and pgbackrest)

**Comment by snopoke** [comment]
If we're using handles it should only restart once (unless it's doing a
reload and a restart)

**Comment by javierwilson** [comment]
true. it restarts only once now, still, it shouldn't restart at all. problem is both plays change the config file, everytime.

**Comment by snopoke** [comment]
Do you have a plan for testing this further?

**Comment by javierwilson** [comment]
this is being tested in softlayer

**Comment by javierwilson** [comment]
@snopoke @gcapalbo any reason not to merge this?

File: ansible/roles/pgbackrest/tasks/main.yml (hunk 1 above)
Comment (snopoke) [reply ansible/roles/pgbackrest/tasks/main.yml]: yea, that's unfortunate

**Comment by gcapalbo** [comment]
@javierwilson this is already deployed to india and enikshay right?

i'm curious if you still think this is a better backup solution. i've noticed some limitations around using pgbackrest, mainly:

- you can only delete the oldest backup, so if you accidentally did a full backup and it's the latest one, there's no way to delete it without deleting all backups.

- you can only delete incremental backups by deleting the full backup they're associated with, so it becomes basically impossible to delete even very old incremental backups. for this reason i had to switch to using rolling differential backups otherwise we ran out of space fast.

- for us, the incremental and differential backups seem to take up just about as much space as the full backups do

- it does allow us to do a point in time recovery for as long we keep the wal archive (which is something we couldn't do before), but keeping wal for a long time takes up too much space so i had to reduce wal retention to the last backup, which means we'd only be able to do the PITR for any time since the last backup

**Comment by javierwilson** [comment]
It doesn't sound as wonderful as it once did :(

File: ansible/roles/postgresql/templates/pg_hba.conf.j2 [hunk 2]
```diff
@@ -99,6 +99,9 @@ host    all             all             ::1/128                 md5
 # Allow replication connections from localhost, by a user with the
 # replication privilege.
 local   replication     postgres                                peer
+{% if hot_standby_server|default(False) %}
+host    replication     postgres        {{ hot_standby_server }}/32 trust
```
Comment (snopoke) [inline ansible/roles/postgresql/templates/pg_hba.conf.j2]: If this get's changed from `trust` to `md5` where would we put the password for pg_backrest?

Comment (snopoke) [reply ansible/roles/postgresql/templates/pg_hba.conf.j2]: Just trying to figure out how we merge this with https://github.com/dimagi/commcarehq-ansible/pull/962/files#diff-d959b4d4814d1bc5d1d095125136a90aR105

Comment (javierwilson) [reply ansible/roles/postgresql/templates/pg_hba.conf.j2]: From the docs "It is possible to configure a password in the primary_conninfo setting but using a .pgpass file is more flexible and secure."
But our current pgbackrest setup doesn't need it, I used this to be able to remotely pg_basebackup, instead of doing it locally and then rsync (saves an step).

Comment (javierwilson) [reply ansible/roles/postgresql/templates/pg_hba.conf.j2]: it's ok to remove this or change to md5

File: ansible/roles/postgresql/tasks/main.yml [hunk 3]
```diff
@@ -60,6 +60,14 @@
 - name: assert hot standby config correct
   assert: { that: "hot_standby_server is not defined or hot_standby_server in groups.pg_standby"}
 
+- name: Create postgresql_log_directory
+  file:
+    dest: "{{ postgresql_log_directory }}"
+    state: directory
+    owner: postgres
+    group: postgres
+  when: postgresql_log_directory|default(false)
```
Comment (snopoke) [inline ansible/roles/postgresql/tasks/main.yml]: why is this necessary? this seems unrelated to pg_backrest

Comment (javierwilson) [reply ansible/roles/postgresql/tasks/main.yml]: following advise on https://github.com/pgbackrest/pgbackrest/issues/413
just to save disk space
